src/rt-app.c

```diff
@@ -415,6 +434,72 @@ static int run_event(event_data_t *event, int dry_run,
 			pthread_yield();
 		}
 		break;
+	case rtapp_fork:
+		{
+			log_debug("fork");
+
+			/*
+			 * If multiple threads race to fork, we must ensure
+			 * each one sees a unique index. Hence the lock.
+			 */
+			pthread_mutex_lock(&fork_mutex);
+			int new_thread_index = nthreads++;
```

Comment:
So main could already have called join on "old" nthreads before fork events happened, no?
In case duration is not set I mean.

Reply:
What's the scenario you're thinking of? I am not sure I understand the problem you're pointing out.

In general, any tasks created by main, pthread_join() will not return until that task has terminated.

For example if we spawn 3 tasks 0, 1 and 2 and task 2 will fork. pthread_join() in main will wait first for 0 to terminate, then 1 then 2. And 2 will not terminate until it has forked all possible tasks. After that pthread_join() will start cycling through all forked tasks with a call to pthread_join().

Assuming we have nested forks; main() will not call pthread_join() on any child thread before its parent has returned/terminated first.

Makes sense? Does this answer your concern somehow?

Reply:
> After that pthread_join() will start cycling through all forked tasks with a call to pthread_join().

Because running_threads is volatile.

Might deserve a comment above the loop in __shutdown() I guess (also because the comment there is at the moment is kind of broken :-).

Reply:
It would help if you're more explicit on how the race is happening and what are you requesting to be changed. Also more info on how the comment is broken and what needs to be done will help me to address it appropriately so that we can finally close this long standing poll request.

Reply:
Argh, sorry I wasn't clear.
I meant that you are right in that we shouldn't have any problem (because running_threads is volatile, I added). Just thought that we might want to enrich the comment above [the loop](https://github.com/scheduler-tools/rt-app/blob/master/src/rt-app.c#L490) using runnning_threads, but no big deal if we don't. :-)

Reply:
Of course. I am more than happy to update it with more info now I understand what you meant :-)

I'll wait to ensure that there are no further comments and will update the PR.

Thanks!

Reply:
```
diff --git a/src/rt-app.c b/src/rt-app.c
index 66b7c31..097d206 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -590,7 +590,16 @@ static void __shutdown(bool force_terminate)
        if (pthread_mutex_trylock(&joining_mutex))
                return;

-       /* wait up all waiting threads */
+       /*
+        * Wait for all threads to terminate.
+        *
+        * pthread_join() will block until the process has terminated. If any
+        * thread forks any processes then it'll update running_threads before
+        * it returns and since running_threads is volatile the for() loop is
+        * guaranteed to check against the updated value in each iteration.
+        * Hence we are guaranteed to wait for all forked threads beside the
+        * originally created ones at startup time.
+        */
        for (i = 0; i < running_threads; i++)
        {
                int ret = pthread_join(threads[i], NULL);
```

I'll fold this up to the "Add a new 'fork' event" commit.

Reply:
Nice, thanks!